### DISCUSSION TITLE
Ability to set `cmd = "path/to/metals-bin"`

### DISCUSSION THREAD

**Discussion by gvolpe** [scalameta/nvim-metals#454] [Q&A] [answered]
Hi Chris,

I'd like to add `nvim-metals` support to my [neovim-flake](https://github.com/gvolpe/neovim-flake) project.

I was wondering if there is something like `lspconfig.metals.setup { cmd = "path/to/metals-bin"; }`? The `useGlobalExecutable` option requires a global executable, which I'd like to avoid.

Thanks!

**Comment by ckipp01** [comment, marked as answer]
Hey @gvolpe! Thanks for the question!

The really short answer is _sure_, you can do the following:

```lua
local metals_config = require("metals").bare_config()
metals_config.cmd = { "whatever/you/want" }
```

However I don't actually document it. The reason for this is that `nvim-metals` actually also takes into account `JAVA_OPTS`, `JAVA_FLAGS`, your `.jvmopts` file and also anything you pass into `metals_config.settings.serverProperties`. It also then strips out certain ones that won't work with metals/coursier automatically and then passes them all in to Neovim to start the server. So if you override `cmd` you lose all of that, which is for sure not what most people want.

There is a `useGlobalExecutable` setting that is a boolean setting to use a global executable, but from what I understand this isn't desired. What we _could_ do is make that setting a `boolean|string` and you can either turn it on or pass in a path where Metals is installed. Then it would still take all the jvm opts stuff into account, but still allow you to pass in a path. Would that work for your use case?

**Reply by gvolpe** [reply]
Thansk @ckipp01 ! 

I think losing all the environment variables stuff would be acceptable for people using Nix, as most would prefer a reproducible way to get those settings in. 

But I'd say it would be nice if we could override `metals_config.cmd` while keeping `metals_config.settings.serverProperties`, I understand that wouldn't be the case now?

WRT `useGlobalExecutable` as `boolean|string`, I think that would be quite confusing, given the naming of that variable. If that's the road to get there, I'd vote for a new `metalsBinaryPath` or so that would take precedence.

**Reply by ckipp01** [reply]
https://github.com/scalameta/nvim-metals/pull/455
